Use both cores of GitHub runners

tautschnig opened the pull request:
According to
https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources,
all GitHub runners feature two CPU cores. Let's use them, as we used to
do in Travis and CodeBuild configurations.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented:
Because most of our time is spent in regression tests and some regression test suites take longer than others, rather than passing `-j` to make or ctest for regression testing I’d recommend install `Thread::Pool::Simple` and setting `TESTPL_JOBS=2` before running the tests, this will make the actual tests run in parallel instead. `cbmc` tests take especially long so I’d expect a greater speedup that way.

NlightNFotis reviewed (approved):
Excellent. :+1:

How much of a speedup do we get with this? Last time I remember builds where taking between 50mins-1hour across all platforms. Hopefully this drops it to about 30-40mins.

thomasspriggs commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -52,22 +51,21 @@ jobs:
           DEBIAN_FRONTEND: noninteractive
         run: |
-          sudo apt-get install -yq cmake ninja-build gcc g++ maven flex bison libxml2-utils cpanminus dpkg-dev
-          cpanm Thread::Pool::Simple
```
Why is this perl module being removed? It is required for the multi threading functionality of `test.pl`. This should currently be enabled via the `TESTPL_JOBS` environment variable.

NlightNFotis replied:
Yes, sorry, I overlooked it - please don't delete this, this is actually helpful.

tautschnig replied:
Except ... we never set TESTPL_JOBS anywhere? I indeed was pondering to set this variable, but then we had been successful on Travis in the past with the use of GNU parallel, hence I opted for using the `test-parallel` target instead. But I'm fine either way. FWIW, the time for running the regression tests was almost exactly halved.

NlightNFotis replied:
`TESTPL_JOBS` at least for this particular action, is being set in line 71 of this file as an environment variable, to my understanding.

thomasspriggs replied:
I think the `test.pl` level of multi threading is included in the existing `.yaml` but isn't actually working correctly. Looking at the CI log from another PR there are existing warnings.
<details>
<summary> `cpanm` output </summary>

```
!
! Can't write to /usr/local/share/perl/5.30.0 and /usr/local/bin: Installing modules to /home/runner/perl5
! To turn off this warning, you have to do one of the following:
!   - run me as a root or with --sudo option (to install to /usr/local/share/perl/5.30.0 and /usr/local/bin)
!   - Configure local::lib in your existing shell to set PERL_MM_OPT etc.
!   - Install local::lib by running the following commands
!
!         cpanm --local-lib=~/perl5 local::lib && eval $(perl -I ~/perl5/lib/perl5/ -Mlocal::lib)
!
--> Working on Thread::Pool::Simple
Fetching http://www.cpan.org/authors/id/J/JW/JWU/Thread-Pool-Simple/Thread-Pool-Simple-0.25.tar.gz ... OK
Configuring Thread-Pool-Simple-0.25 ... OK
Building and testing Thread-Pool-Simple-0.25 ... OK
Successfully installed Thread-Pool-Simple-0.25
```

</details>  


<details>
<summary> `test.pl` output </summary>

```
1: Jobs set but thread pool module not found,
1: install with 'cpan -i Thread::Pool::Simple'
```

</details> 

I am fine with removing this, as it isn't currently working. However I would prefer the removal to be in a separate commit with the "because it doesn't work" explanation. That way if we ever want to re-instate it we can revert that commit only.

tautschnig replied:
Thanks, will move this to a commit of its own!

thomasspriggs reviewed (commented):
Does CI actually finish in less time with these changes?

tautschnig commented:
> Excellent. 👍
> 
> How much of a speedup do we get with this? Last time I remember builds where taking between 50mins-1hour across all platforms. Hopefully this drops it to about 30-40mins.

I have added the info to the commit message: on Ubuntu it's down to 30-40 minutes, and MacOS it's under 1 hour. Windows I didn't yet have data for #5606 had not been merged yet.

tautschnig commented:
> Because most of our time is spent in regression tests and some regression test suites take longer than others, rather than passing `-j` to make or ctest for regression testing I’d recommend install `Thread::Pool::Simple` and setting `TESTPL_JOBS=2` before running the tests, this will make the actual tests run in parallel instead. `cbmc` tests take especially long so I’d expect a greater speedup that way.

As said in https://github.com/diffblue/cbmc/pull/5608#discussion_r525086706, I'd be happy to go down that route. I just picked the somewhat more conservative approach that seemed to work fine in the past.

thomasspriggs commented on .github/workflows/pull-request-checks.yaml:
```diff
@@ -52,22 +51,21 @@ jobs:
-        run: cd build; ctest . -V -L CORE
+        run: cd build; ctest . -V -L CORE -j2
         env:
           TESTPL_JOBS: 2
```
:no_entry_sign: If we are removing the installation of `Thread::Pool::Simple` we should also remove the setting of this environment variable. That will keep things consistent and avoid `test.pl` attempting to use a missing module.

thomasspriggs reviewed (changes requested):
I intend to approve once the issue of partial removal of the `test.pl` level multi-threading support is resolved.

tautschnig committed:
GitHub workflow: Don't unnecessarily install Thread::Pool::Simple

It doesn't actually work, as discussed in
https://github.com/diffblue/cbmc/pull/5608#discussion_r525107486.

tautschnig committed:
Use both cores of GitHub runners

According to
https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources,
all GitHub runners feature two CPU cores. Let's use them, as we used to
do in Travis and CodeBuild configurations. This reduces the build time
on Ubuntu runners from 50+ minutes to 30-40 minutes, and on MacOS from
1.5 hours to under 1 hour.

For Windows, work around a CMake issue discussed in
https://gitlab.kitware.com/cmake/cmake/-/issues/20564.

tautschnig commented:
> I intend to approve once the issue of partial removal of the `test.pl` level multi-threading support is resolved.

Should now be done.

thomasspriggs reviewed (approved):
:+1:
